Add a DirectedGossip struct

tomaka opened the pull request:
Provides the primitive that will make it possible to tackle https://github.com/paritytech/polkadot/issues/1453

Adds a new `gossip` module in `sc-network`, containing `DirectedGossip`.
The way it works is what is described [here](https://github.com/paritytech/substrate/blob/6384b85d0b25a99e01cfb62a81a551d27aa5c74b/client/network/src/service.rs#L680-L688).

Here's how you would plug it into Polkadot, regarding https://github.com/paritytech/polkadot/issues/1453:

- When the network bridge learns about a new connection, it builds a new `DirectedGossipPrototype`. This type doesn't have any template parameter.
- The [`NetworkBridgeEvent::PeerConnected`](https://github.com/paritytech/polkadot/blob/cc2d7afd128b259791139359067aa519eadbf205/node/subsystem/src/messages.rs#L140) enum variant should contain a new field of type `DirectedGossipPrototype` (or some abstraction of it). The bridge passes the newly-created object there.
- Whatever subsystem picks up the `PeerConnected` event turns the prototype into an actual `DirectedGossip`, passing the protocol name, queue size, and messages type.
- The subsystem can then manipulate the queue of notifications using `push_or_discard`, `push_unbounded`, or `retain`.
- When a `PeerDisconnected` message is received, the subsystem discards the `DirectedGossip` for this peer.

tomaka commented:
I marked as "in-progress" because tests have to be written.

rphmeier commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Helper for sending rate-limited gossip messages.
```
I feel like `Gossip` is a bit of a misnomer here

twittner replied:
I concur. Renaming `DirectedGossip` to `Sender` would sound better to me.

rphmeier commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+//!
+//! If multiple instances of [`DirectedGossip`] exist for the same peer and protocol, or if some
+//! other code uses the [`NetworkService`] to send notifications to this peer and protocol, then
+//! the notifications will be interleaved in an unpredictable way.
```
I think it's implied, but can we state explicitly that notifications coming from the same `DirectedGossip` instance will always (or almost always) be delivered in order, even if there are other instances or users of `write_notification`?

twittner commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+impl<'a, M> Drop for QueueLock<'a, M> {
+	fn drop(&mut self) {
+		// We notify the `Condvar` in the destructor in order to be able to push multiple
+		// messages and wake up the background task only one afterwards.
```
```suggestion
		// messages and wake up the background task only once afterwards.
```

twittner commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+/// Abstraction around `NetworkService` that permits removing the `B` and `H` parameters.
+trait AbstractNotificationSender {
```
I suggest to remove this trait. The two type parameters `B` and `H` are only necessary in the private `spawn_task` method and adding those is far less code than this trait plus impl. And your calls to `NetworkService::notification_sender` are not dynamically dispatched.

tomaka replied:
It's necessary for `DirectedGossipPrototype` to work. See your other comment.

twittner commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+#[derive(Debug)]
+struct SharedLock<M> {
```
Unless you plan to add more to `SharedLock` I suggest to remove the type and use the `Mutex` directly around the `VecDeque`. If you want to keep it I think a name such as `Queue` might be more apt.

twittner commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+/// Utility. Generic over the type of the messages. Holds a [`NetworkService`] and a [`PeerId`].
+/// Provides a [`DirectedGossipPrototype::build`] function that builds a [`DirectedGossip`].
+#[derive(Clone)]
+pub struct DirectedGossipPrototype {
```
This is `pub` so I assume this will be used from somewhere external? Could you explain this a little more please? Currently it is not in use and the `DirectedGossip` and task construction could happen directly in `DirectedGossip::new`.

tomaka replied:
I admit that this is type is specifically targeted for the Polkadot use case.

In Polkadot's code, the so-called network bridge communicates [via messages](https://github.com/paritytech/polkadot/blob/cc2d7afd128b259791139359067aa519eadbf205/node/subsystem/src/messages.rs#L138-L153) with other subsystems. For example, when we connect, a `PeerConnected` message is sent.

The network bridge doesn't know what is the type of the messages that the various subsystems would manipulate (the `M` generics), and the various subsystems don't know what the `B` and `H` generics would be.

This `DirectedGossipPrototype` would easily plug into that scheme.

twittner replied:
Should Polkadot-specific code not be put into Polkadot and not in here?

rphmeier replied:
> The network bridge doesn't know what is the type of the networking messages that the various subsystems would manipulate (the M generic in this PR), and the various subsystems don't know what the B and H generics would be.

I am actually changing this right now, so we may not need this type. Will push a PR shortly

rphmeier replied:
https://github.com/paritytech/polkadot/pull/1535 - now the network bridge is aware of the specific information flowing over the network and we may be able to avoid these prototypes. However, different subsystems will still want `Sender`s for different types. However, the sender for a specific variant should also be encoding as `|x| ValidationProtocolV1::SomeConcreteVariant(x).encode()` which I think is covered by the encode-fn part of the API here

tomaka replied:
> Should Polkadot-specific code not be put in Polkadot?

To answer specifically this: the way I see it, this code isn't really Polkadot-specific.
This code provides a convenient-to-use high-level API on top of a low-level primitive, and people are free to use either the convenient API if it suits them, or the low-level primitive if it doesn't suit them.

I think that, in general, we want to be as "universal" as possible when designing low-level code, but as long as the lower-level primitives are exposed, higher-level code can be a bit more targeted towards certain use-cases.

tomaka commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+		// being asynchronous, it can't reasonably be locked from within a destructor.
+		// For this reason, this destructor is a "best effort" destructor.
+		// See also the corresponding code in the background task.
+		self.shared.stop_task.store(true, atomic::Ordering::Acquire);
```
It's maybe the 10th time I try, but I still don't understand anything about `Ordering`

rphmeier replied:
`Acquire` and `Release` are pretty similar but `Acquire` is for reads and `Release` is for writes. I think `store, Acquire` will panic at runtime

tomaka commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,318 @@
+
+	/// Pushes a message to the queue. Does not enforce any limit to the size of the queue.
+	/// Use this method only if the message is extremely important.
+	pub fn push_unbounded(&mut self, message: M) {
```
I'll remove it, as it it seems that if you need this function then you probably have an XY problem. We can indeed add it back later.

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+	queue_size_limit: usize,
+}
+
+impl<'a, M: Send + 'static> QueueLock<'a, M> {
```
Do I understand correctly that a user has no way to register its underlying Futures task via its `Context` to be notified once a full queue has an empty slot again? Hence a user can only proactively add items to the queue. Is this something that users might need?

tomaka replied:
It is by design that no waiting is involved anywhere (except shortly for locking the mutex in the situation where the background task happens to be popping an element right at this moment).

When you want to notify your peers of something, you push an element to the buffer. This buffer is what keeps track of what has been sent or not. This buffer should never be full unless the peer is completely unresponsive (in which case it is guaranteed that a `NotificationsSubstreamClosed` event will be generated).

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+/// Locked queue of messages to the given peer.
+///
+/// As long as this struct exists, the background task is asleep and the owner of the [`QueueLock`]
+/// is in total control of the buffer.
```
How about adding a comment that sending is only happening once this struct is dropped?

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+		// For this reason, this destructor is a "best effort" destructor.
+		// See also the corresponding code in the background task.
+		self.shared.stop_task.store(true, atomic::Ordering::Release);
+		self.shared.condvar.notify_all();
```
Do I understand correctly that at any given point in time, there can only be a single task waiting for this condvar? If so, why use `notify_all` here? I am fine with using `notify_all` everywhere to prevent deadlocks introduced with future changes.

mxinden replied:
To make sure my understanding is correct, why would we ever have multiple background tasks for a single `Shared` struct?

tomaka replied:
We'll probably never run multiple tasks, but also I don't see a preference for using `notify_one` either.

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+		// The "clean" way to notify the `Condvar` here is normally to first lock the `Mutex`,
+		// then notify the `Condvar` while the `Mutex` is locked. Unfortunately, the `Mutex`
+		// being asynchronous, it can't reasonably be locked from within a destructor.
+		// For this reason, this destructor is a "best effort" destructor.
```
This is not so much a "best effort" but rather a "maybe delayed" destructor, right?

tomaka commented on client/network/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ targets = ["x86_64-unknown-linux-gnu"]
 prost-build = "0.6.1"
 
 [dependencies]
+async-std = { version = "1.6.2", features = ["unstable"] }
```
No, it is for `Convar`. It is unfortunate that we have to depend on the `unstable` feature, but I couldn't find any crate other than `async-std` that provides an asynchronous `Condvar`.

mxinden replied:
If this is something we want to avoid, one can use a channel as a condvar.

```rust
use futures::channel::mpsc::channel;

fn main() {
    let (tx, rx) = channel(0);
    
    let producer = async move {
        // Produce something and put it to place X.
        
        // Signal the consumer that that something can be optained.
        tx.send(()).await;
    }
    
    let consumer = async move {
        // Wait for the signal. Maybe include a timeout like you do today with 10 secs.
        let _ = rx.next().await;
        
        // Access place X.
    }
}
```

Problem is that this can't be used in the `Drop` implementations and thus cleanup would depend on the 10 sec timeout.

twittner replied:
> Problem is that this can't be used in the Drop implementations [...]

It does not always need to. When the `Sender` is dropped, the receiver will notice and the task can terminate. Instead of `let _ = rx.next().await;` one would write `if rx.next().await.is_some() { ... }`. In the Drop impl of QueueGuard one can use `Sender::try_send`.

tomaka replied:
I find it a bit weird to use a channel, which involves an additional `Arc`, `Mutex` and `Vec`, just to wake up a task, rather than a `Waker`.

I have now also tried using a `Waker`, but the implementation is considerably more tricky because of potential race conditions and having to introduce manual polling within an `async` function and having to implement your own `Waker`.

I don't really understand why we don't want the `Condvar` solution, as a `Condvar` is exactly the tool that is designed for this specific job.

rphmeier commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+	/// Pushes a message to the queue, or discards it if the queue is full.
+	///
+	/// The returned `Future` is expected to be ready quite quickly.
+	pub async fn queue_or_discard(&self, message: M)
```
This API could return a `Result<(), M>` or similar as well

tomaka replied:
It's a convenient shortcut.

rphmeier commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,312 @@
+
+impl<'a, M: Send + 'static> QueueLock<'a, M> {
+	/// Pushes a message to the queue, or discards it if the queue is full.
+	pub fn push_or_discard(&mut self, message: M) {
```
Same here. It would be good to get back the `message` if the queue is full

tomaka replied:
But what would you do with the message that is returned? Put it in another queue?
The only sane thing you can do when the queue is full is either discard the message or force-close the connection.

Additionally, what if the connection with the remote is closed? Are we supposed to return back the message as well? If so, then it's very problematic because we can't detect this.

rphmeier replied:
> The only sane things you can do when the queue is full is either discard the message or force-close the connection.

Isn't `retain` another option? We could decide to do that based on the type of the message we are trying to send.

And is there no sane thing you can do? Why can't you wait for space to appear in the queue or something like that?

tomaka replied:
The way I see `retain` is that you'd call it all the time, even if there is space in the buffer. As far as I can tell, when a message is obsolete, there is no point in leaving it in the queue anyway.

> Why can't you wait for space to appear in the queue or something like that?

The entire reason for this API to exist is to remove the need for any waiting. See also [this paragraph](https://github.com/paritytech/substrate/blob/270a992bf299cd41b2ed4e211814d2a958e3bfc8/client/network/src/service.rs#L653-L671).

Ultimately there *has to* be a code somewhere that holds some sort of `HashMap<PeerId, DirectedGossip>`. If it needs to send a message to one of the peers and its buffer is full, then it shouldn't wait for this peer and instead continue its processing.

rphmeier replied:
It would be nice to have a way to wait, although I agree that the API for that should be used sparingly so you don't degrade to the performance of the slowest peer. There are cases where we don't want to drop messages, for instance when responding to a validator's request.

tomaka replied:
> There are cases where we don't want to drop messages

I believe that everything that would fall in this category should be covered by request-response protocols.

Rather than adding a wait, I could restore [the `push_unbounded` method that I've removed after a review](https://github.com/paritytech/substrate/pull/6803/commits/c11bcfb5cecd983085eacf3f2b08f9a6e27c6651#diff-f5e0015f0355ca80bb672b3980b8eba1L218-L223).

mxinden replied:
@rphmeier @tomaka is this conversation resolved?

tomaka commented:
I have addressed all concerns, except for the `QueueSenderPrototype` (formerly `DirectedGossipPrototype`) that might be useless and add a very small overhead.

@rphmeier Do you confirm that the `NetworkBridge` would directly create the `QueueSender`s itself? In other words, that it knows what the `M` generic is, and that there's no need for some "prototype" anymore?
If so I'll remove it.

rphmeier commented:
> Do you confirm that the NetworkBridge would directly create the QueueSenders itself? In other words, that it knows what the M generic is, and that there's no need for some "prototype" anymore?

Yes, after https://github.com/paritytech/polkadot/pull/1537 is in

tomaka commented:
bot rebase

twittner commented:
Any comment on @mxinden's [suggestion](https://github.com/paritytech/substrate/pull/6803#discussion_r466192207) to use a channel instead of a `Condvar` which would avoid depending on `unstable` features from `async-std`?

tomaka commented:
If we do indeed want to not depend on `unstable`, then I'd suggest having a `Waker` instead of a channel.
However, the `Mutex`/`Condvar` "pattern" is what you're supposed to use for that kind of code, so do we actually want to get rid of the `Condvar`?

tomaka commented:
bot rebase

tomaka commented:
Friendly ping

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,245 @@
+mod tests;
+
+/// Notifications sender for a specific combination of network service, peer, and protocol.
+pub struct QueueSender<M> {
```
```suggestion
pub struct QueuedSender<M> {
```

Not sure whether this is a typo or intentional. In case it is the former I would prefer `QueuedSender`.

mxinden commented on client/network/src/gossip.rs:
```diff
@@ -0,0 +1,245 @@
+/// is in total control of the buffer. Messages can only ever be sent out after the [`QueueGuard`]
+/// is dropped.
+#[must_use]
+pub struct QueueGuard<'a, M> {
```
:+1:

rphmeier commented:
> @rphmeier @tomaka is this conversation resolved

On discarding, I will defer to Pierre for now. If we find a need, the behavior can be easily changed later.

mxinden commented on client/network/src/gossip/tests.rs:
```diff
@@ -0,0 +1,201 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{config, gossip::QueueSender, Event, NetworkService, NetworkWorker};
```
```suggestion
use crate::{config, gossip::QueuedSender, Event, NetworkService, NetworkWorker};
```
Hence the CI failures.

mxinden commented on client/network/src/gossip/tests.rs:
```diff
@@ -0,0 +1,201 @@
+
+	async_std::task::block_on(async move {
+		let (sender, bg_future) =
+			QueueSender::new(node1, node2_id, ENGINE_ID, NUM_NOTIFS, |msg| msg);
```
```suggestion
			QueuedSender::new(node1, node2_id, ENGINE_ID, NUM_NOTIFS, |msg| msg);
```

tomaka commented:
bot merge
